runtime/criusupport/criusupport.cpp

```diff
@@ -312,28 +312,45 @@ toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend, BOOLEAN t
 	}
 }
 
+#define J9VM_DELAYCHECKPOINT_CLINIT ((void*)0x1)
+#define J9VM_DELAYCHECKPOINT_NOTCHECKPOINTSAFE ((void*)0x2)
+
 static UDATA
-notCheckpointSafeFrameWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState)
+notCheckpointSafeOrClinitFrameWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState)
 {
 	J9Method *method = walkState->method;
+	UDATA returnCode = J9_STACKWALK_KEEP_ITERATING;
+
 	if (NULL != method) {
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
 		J9ClassLoader *methodLoader = J9_CLASS_FROM_METHOD(method)->classLoader;
+		J9UTF8 *romMethodName = J9ROMMETHOD_NAME(romMethod);
+		if (J9UTF8_LITERAL_EQUALS(J9UTF8_DATA(romMethodName), J9UTF8_LENGTH(romMethodName), "<clinit>")) {
+			walkState->userData4 = J9VM_DELAYCHECKPOINT_CLINIT;
```

Comment:
> I think there should be an (experimental) option to enable / disable this behaviour to make it easier to experiment with if users find their checkpoints are failing due to .

That can be added.

>  "Delayed" would attempt to let complete but still eventually do the checkpoint after some # of retries

The fundamental issue is that if the class causing the delay is used in a checkpoint/restore hook, it will cause a deadlock. 

> Is there a design / issue that describes the rationale for why the current approach treats them the same?

The issue is https://github.com/eclipse-openj9/openj9/issues/15806. This comment, https://github.com/eclipse-openj9/openj9/issues/15806#issuecomment-1538636409, shows the deadlock.

I havent added more details to this issue as I am still verifying that this solution works.